parsely/src/main/java/com/parsely/parselyandroid/ParselyTracker.kt

```diff
@@ -0,0 +1,424 @@
+            }
+        }
+        flushQueue = FlushQueue(
+            flushManager,
+            localStorageRepository,
+            ParselyAPIConnection(ROOT_URL + "mobileproxy"),
+            sdkScope,
+            AndroidConnectivityStatusProvider(context)
+        )
+        intervalCalculator = HeartbeatIntervalCalculator(clock)
+
+        isDebug = false
+        flushManager.start()
+        ProcessLifecycleOwner.get().lifecycle.addObserver(
+            LifecycleEventObserver { _, event: Lifecycle.Event ->
+                if (event == Lifecycle.Event.ON_STOP) {
+                    flushEvents()
+                }
+            }
+        )
+    }
+
+    /**
+     * Get the heartbeat interval
+     *
+     * @return The base engagement tracking interval.
+     */
+    public val engagementInterval: Double?
+        get() = engagementManager?.intervalMillis
+
+    public val videoEngagementInterval: Double?
+        get() = videoEngagementManager?.intervalMillis
+
+    /**
+     * Returns whether the engagement tracker is running.
+     *
+     * @return Whether the engagement tracker is running.
+     */
+    public fun engagementIsActive(): Boolean {
+        return engagementManager?.isRunning ?: false
+    }
+
+    /**
+     * Returns whether video tracking is active.
+     *
+     * @return Whether video tracking is active.
+     */
+    public fun videoIsActive(): Boolean {
+        return videoEngagementManager?.isRunning ?: false
+    }
+
+    /**
+     * Returns the interval at which the event queue is flushed to Parse.ly.
+     *
+     * @return The interval at which the event queue is flushed to Parse.ly.
+     */
+    public val flushInterval: Long
+        get() = flushManager.intervalMillis / 1000
+
+    /**
+     * Set a debug flag which will prevent data from being sent to Parse.ly
+     *
+     *
+     * Use this flag when developing to prevent the SDK from actually sending requests
+     * to Parse.ly servers. The value it would otherwise send is logged to the console.
+     *
+     * @param debug Value to use for debug flag.
+     */
+    public fun setDebug(debug: Boolean) {
+        isDebug = debug
+        log("Debugging is now set to $isDebug")
+    }
+
+    /**
+     * Register a pageview event using a URL and optional metadata.
+     *
+     * @param url         The URL of the article being tracked
+     * (eg: "http://example.com/some-old/article.html")
+     * @param urlRef      Referrer URL associated with this video view.
+     * @param urlMetadata Optional metadata for the URL -- not used in most cases. Only needed
+     * when `url` isn't accessible over the Internet (i.e. app-only
+     * content). Do not use this for **content also hosted on** URLs Parse.ly
+     * would normally crawl.
+     * @param extraData   A Map of additional information to send with the event.
+     */
+    public fun trackPageview(
+        url: String,
+        urlRef: String? = null,
```

Comment:
```suggestion
        urlRef: String = "",
```

I suggest that we update all the `urlRef` parameters to use this ^ instead of the nullable value and remove the `orEmpty()`.

This may seem like a nitpick, but besides the simplification, it exposes what we are doing with this argument in case it's omitted, so there is some value in it. I guess it's fairly obvious what we'd be doing with a `null` value in this case, but as a general principle, I think it's a good idea to be transparent when we can.

Reply:
Good point, I agree and addressed in ea37072. Initially, I was super cautious about breaking public API, but on the other hand, now is the best moment to do so (we'll be releasing a new major version anyway).

Reply:
I think there are still a few more `urlRef`s that use the nullable approach.

Reply:
Thank you, missed it completely 🤦 . Fixed in 697c051